alf/networks/mdq_critic_networks.py

```diff
@@ -0,0 +1,573 @@
+            action_qt = ActionQuantizer(action_spec, "uniform", 15)
+        self._action_qt = action_qt
+        self._action_bins = self._action_qt._action_bins
+
+        # the logpi of the uniform prior used for KL computation
+        self._log_pi_uniform_prior = -np.log(self._action_bins)
+
+        self._action_dim = action_spec.shape[0]  # control vector dim
+        self._num_critic_replicas = num_critic_replicas
+
+        self._obs_encoding_net = ParallelEncodingNetwork(
+            observation_spec,
+            self._num_critic_replicas,
+            fc_layer_params=obs_encoding_layer_params,
+            activation=activation,
+            kernel_initializer=kernel_initializer)
+
+        last_activation = math_ops.identity
+        last_kernel_initializer = functools.partial(torch.nn.init.uniform_, \
+                                a=-0.003, b=0.003)
+
+        in_size = self._action_dim
+
+        self._pre_encoding_nets = nn.ModuleList()
+        for i in range(self._action_dim):
+            # output_spec.shape: [n, d]
+            self._pre_encoding_nets.append(
+                ParallelEncodingNetwork(
+                    TensorSpec((self._obs_encoding_net.output_spec.shape[-1] +
+                                in_size, )),
+                    self._num_critic_replicas,
+                    fc_layer_params=pre_encoding_layer_params,
+                    activation=activation,
+                    kernel_initializer=kernel_initializer))
+
+        # parallel along both critic and action dims without sharing parameters
+        # for each action dimension.
+        # input: [B, action_dim*n, d]: need to stack over dim1
+        # output: [B, action_dim*n, d']: need to unstack over dim1 for
+        # splitting over networks
+        self._pre_encoding_parallel_net = ParallelEncodingNetwork(
+            TensorSpec(
+                (self._obs_encoding_net.output_spec.shape[-1] + in_size, )),
+            self._num_critic_replicas * self._action_dim,
+            fc_layer_params=pre_encoding_layer_params,
+            activation=activation,
+            kernel_initializer=kernel_initializer)
+
+        # parallel along both critic and action dims with sharing parameters
+        # for each action dimension.
+        # input: [action_dim*B, n, d]: need to stack over dim0
+        # output: [action_dim*B, n, d']: need to unstack over dim0 for
+        # splitting over networks
+        self._mid_shared_encoding_nets = ParallelEncodingNetwork(
+            TensorSpec(
+                (self._pre_encoding_parallel_net.output_spec.shape[-1], )),
+            self._num_critic_replicas,
+            fc_layer_params=mid_encoding_layer_params,
+            activation=activation,
+            kernel_initializer=kernel_initializer)
+        out_size = self._mid_shared_encoding_nets.output_spec.shape[-1]
+
+        post_enc_out_size = self._action_qt.action_bins
+
+        self._post_encoding_nets = nn.ModuleList()
```

Comment:
since post_encoding_nets will not be trained. It's better use a list instead of ModuleList to hold it. Same for pre_encoding_nets.

Reply:
Good point. Switched to list.